crates/iceberg/src/spec/values.rs

```diff
@@ -3604,4 +3608,29 @@ mod tests {
 
         assert_eq!(result, expected);
     }
+
+    #[test]
+    fn test_record_ser_de() {
+        let expected_literal = Literal::Struct(Struct::from_iter(vec![
+            Some(Literal::Primitive(PrimitiveLiteral::Int(1))),
+            Some(Literal::Primitive(PrimitiveLiteral::String(
+                "bar".to_string(),
+            ))),
+            None,
+            Some(Literal::Primitive(PrimitiveLiteral::Int(1000))),
+        ]));
+        let fields = Type::Struct(StructType::new(vec![
+            NestedField::required(2, "id", Type::Primitive(PrimitiveType::Int)).into(),
+            NestedField::optional(3, "name", Type::Primitive(PrimitiveType::String)).into(),
+            NestedField::optional(4, "address", Type::Primitive(PrimitiveType::String)).into(),
+            NestedField::required(5, "extra", Type::Primitive(PrimitiveType::Int)).into(),
+        ]));
+
+        let raw_literal = RawLiteral::try_from(expected_literal.clone(), &fields).unwrap();
+        let serialized = serde_json::to_string(&raw_literal).unwrap();
+        let deserialized: RawLiteral = serde_json::from_str(&serialized).unwrap();
+        let deserialized_literal = deserialized.try_into(&fields).unwrap().unwrap();
+
+        assert_eq!(expected_literal, deserialized_literal);
```

Comment:
It looks like the whole serialization is off, this should be [done by ID](https://iceberg.apache.org/spec/?column-projection#json-single-value-serialization) instead of name:

I checked out the branch, and it is currently by name:
```json
{
    "id": 1,
    "extra": 1000,
    "name": "bar",
    "address": null
}
```

While it should be:
```json
{
    "2": 1,
    "3": "bar",
    "4": null,
    "5": 1000
}
```

Reply:
This is not same as JSON single-value serialization, th JSON single-value serialization has the specific implementation https://github.com/apache/iceberg-rust/blob/813c2b5610490886e819be97fb5fbf173ad68a58/crates/iceberg/src/spec/values.rs#L1993.
This test case is just to test the normal Serialize implementation, internally it mainly used in avro format. See https://docs.rs/avro-rs/latest/avro_rs/types/struct.Record.html, that's why here record store name and value. 

Here I serialize it into json type is to test the reorder case.

Reply:
Mark unresolve for my newly added comment: https://github.com/apache/iceberg-rust/pull/795#discussion_r1887077718

Sorry @Fokko :smile:

Reply:
Ah, around the same time! I share your concern, and I would like to check later on if we do the field-ID projection properly, but I didn't want to block the release 👍 

For Avro it is very simple, it will always be decoded in the same order as the schema (otherwise it will just break). That said, we can rely on the order for V1, but use field-ID-based projection for V2 tables.